cmd/container-use/stdio_test.go

```diff
@@ -0,0 +1,359 @@
+	require.NoError(t, err)
+	defer os.RemoveAll(sharedConfigDir)
+
+	// Start multiple servers pointing to the same repo
+	servers := make([]*MCPServerProcess, numServers)
+
+	for i := range numServers {
+		ctx := context.Background()
+		containerUseBinary := getContainerUseBinary(t)
+		cmd := exec.CommandContext(ctx, containerUseBinary, "stdio")
+		cmd.Dir = sharedRepoDir
+		cmd.Env = append(os.Environ(), fmt.Sprintf("CONTAINER_USE_CONFIG_DIR=%s", sharedConfigDir))
+
+		mcpClient, err := client.NewStdioMCPClient(containerUseBinary, cmd.Env, "stdio")
+		require.NoError(t, err)
+
+		initRequest := mcp.InitializeRequest{}
+		initRequest.Params.ProtocolVersion = mcp.LATEST_PROTOCOL_VERSION
+		initRequest.Params.ClientInfo = mcp.Implementation{
+			Name:    fmt.Sprintf("Shared Repo Test Client %d", i),
+			Version: "1.0.0",
+		}
+		initRequest.Params.Capabilities = mcp.ClientCapabilities{}
+
+		serverInfo, err := mcpClient.Initialize(ctx, initRequest)
+		require.NoError(t, err)
+
+		servers[i] = &MCPServerProcess{
+			cmd:        cmd,
+			client:     mcpClient,
+			repoDir:    sharedRepoDir,
+			configDir:  sharedConfigDir,
+			serverInfo: serverInfo,
+			t:          t,
+		}
+
+		t.Cleanup(func() {
+			servers[i].Close()
+		})
+	}
+
+	// Perform comprehensive workflow in parallel: create environments, write files, run commands
+	var wg sync.WaitGroup
+	envIDs := make([]string, numServers)
+	errors := make([]error, numServers)
+
+	for i := range numServers {
+		wg.Add(1)
+		go func(serverIdx int) {
+			defer wg.Done()
+
+			// Step 1: Create environment
+			envID, err := servers[serverIdx].CreateEnvironment(
+				fmt.Sprintf("Shared Repo Test %d", serverIdx),
+				fmt.Sprintf("Testing shared repository contention %d", serverIdx),
+			)
+			if err != nil {
+				errors[serverIdx] = fmt.Errorf("environment creation failed: %w", err)
+				return
+			}
+			envIDs[serverIdx] = envID
+
+			// Step 2: Write multiple files
+			for j := range 3 {
+				err := servers[serverIdx].FileWrite(
+					envID,
+					fmt.Sprintf("server%d_file%d.txt", serverIdx, j),
+					fmt.Sprintf("Content from server %d, file %d\nTimestamp: concurrent test", serverIdx, j),
+					fmt.Sprintf("Writing file %d from server %d", j, serverIdx),
+				)
+				if err != nil {
+					errors[serverIdx] = fmt.Errorf("file write failed: %w", err)
+					return
+				}
+			}
+
+			// Step 3: Read back one of the files to verify
+			content, err := servers[serverIdx].FileRead(envID, fmt.Sprintf("server%d_file0.txt", serverIdx))
+			if err != nil {
+				errors[serverIdx] = fmt.Errorf("file read failed: %w", err)
+				return
+			}
+			if content == "" {
+				errors[serverIdx] = fmt.Errorf("file read returned empty content")
+				return
+			}
+
+			// Step 4: Run commands to verify the environment works
+			listOutput, err := servers[serverIdx].RunCommand(
+				envID,
+				fmt.Sprintf("ls -la server%d_*.txt | wc -l", serverIdx),
+				"Count files created by this server",
```

Comment:
🐛 **Bug (Medium):** Potential panic if listOutput is empty - should check array bounds before accessing index 0

```suggestion
lines := strings.Split(strings.TrimSpace(listOutput), "\n")
			if len(lines) == 0 || lines[0] != "3" {
```